src/Block/Fetch.hs

```diff
@@ -85,14 +85,16 @@ startBlockFetcher blockInfo datumFilter = do
   OgmiosInfo ogmiosPort ogmiosAddress <- ask
   MkOgmiosWorkerMVar envOgmiosWorker <- ask
   env <- Reader.ask
+  canStart <- liftIO newEmptyMVar
```

Comment:
Looks like that is a local variable, which was initialized at each function call. So it shouldn't work like a mutex. So I make an alternative solution here: 
- https://github.com/mlabs-haskell/ogmios-datum-cache/pull/41

but I don't tested it. Do you have a hint how to run it quickly?

Reply:
Yeah, but that solution introduces new global MVar to already messy concurrency. Point of `canStart` is to be set to `()` after we are sure that no other thread is running and current worker's thread (actually `Async`) is already in worker mutex. If another thread is running `Async` is cancelled and `canStart` is never set to `()` so no actual block fetching happens. Is there something wrong in creating it each time inside `startBlockFetcher`?

In addition your solution has time of check/time of use error. You check that MVar is empty and if it's later you put some value there. It will break if multiple requests are sent in roughly the same time.

Reply:
The issue is that you don't know if you can insert new worker in mvar until you try, so you have to start worker and then try to insert it. `canStart` will stop execution of actual logic until we get result of `tryPutMVar`, and if it succeeds it will put `()` into `canStart`, worker will take it and begin execution. If it failed it'll never put `()` in `canStart` so actual block fetching will never start and `Async` will be cancelled.

Reply:
Got it. Nice solution! I failed to understand how it works from the first try.

About the time of check/time of use error. I expect, that `envRunLock` works like a synchronized section. Take `()` from MVar for lock, and put it back for unlock. `takeMVar` should block the thread, if nothing to take. Do I make a mistake (don't have much experience with concurrency in haskell and can expect any magic from laziness)?

So the main differences are:
- (your) run a worker (which can execute network activity) and kill it if we already have one
- (mine) if we don't have a worker, run it (which requires a global var for mutex)

Really don't know which is better (of course, if my solution is correct).

Reply:
Nice. I like your solution more.